# Z-Tests Practice Workbook

**Topics covered:**
- One-sample Z-test (mean)
- One-sample proportion Z-test
- Two-proportion Z-test
- Two-sample Z-test with power analysis
- Multiple testing corrections (Bonferroni & Benjamini–Hochberg)

**Solution structure for each question:**

> Hypotheses → Justify Test Choice → Test Statistic → Critical Value Method → p-value Method → Confidence Interval → Conclusion

In [1]:
import numpy as np
from scipy.stats import norm
import math

---
## 📌 Question 1 : One-Sample Z-Test (Mean)

A factory claims that the mean diameter of its bolts is **10 mm**.

A sample of **50 bolts** is taken with the following statistics:

| Parameter | Value |
|---|---|
| Sample mean (x̄) | 10.3 mm |
| Population std dev (σ) | 1.0 mm |
| Sample size (n) | 50 |
| Significance level (α) | 0.05 |

Perform a **two-tailed one-sample Z-test** to evaluate the factory's claim.  
Also compute the **95% confidence interval** for the mean.

#### Step 1 : Hypotheses

- **H₀:** μ = 10 (factory's claim holds)
- **H₁:** μ ≠ 10 (two-tailed, no assumption on which direction the mean shifts)

#### Step 2 : Why Z-Test?

- Population σ is **known** → use Z-test (unknown σ would require a t-test)
- n = 50 is large enough for the Central Limit Theorem to apply
- Both the critical value and p-value methods will be used to cross-verify

In [2]:
# Q1 - given values
xbar1 = 10.3   # sample mean (mm)
mu01  = 10     # hypothesized population mean (mm)
sig1  = 1      # known population std dev (mm)
n1    = 50     # sample size
a1    = 0.05   # significance level

#### Step 3 : Test Statistic

$$z = \frac{\bar{x} - \mu_0}{\sigma / \sqrt{n}}$$

Measures how many standard errors the sample mean is from the claimed population mean.

In [3]:
z1 = (xbar1 - mu01) / (sig1 / n1**0.5)   # z-statistic
z1

2.1213203435596477

#### Step 4 : Critical Value Method

For α = 0.05 (two-tailed), each tail gets α/2 = 0.025.  
We need z such that P(Z < z) = 0.975 → `norm.ppf(0.975)`.

**Decision rule:** Reject H₀ if |z_stat| > z_critical

In [4]:
zcrit1 = norm.ppf(1 - a1/2)   # two-tailed critical value
zcrit1

np.float64(1.959963984540054)

**Decision:** |z| = 2.12 > z_critical = 1.96 → **Reject H₀**

#### Step 5 : p-value Method

$$p = 2 \times P(Z > |z|)$$

`norm.sf(x)` returns P(Z > x), the survival function. Multiply by 2 for a two-tailed test.

In [5]:
pval1 = 2 * norm.sf(abs(z1))   # two-tailed p-value
pval1

np.float64(0.033894853524688816)

**Decision:** p-value ≈ 0.034 < α = 0.05 → **Reject H₀** ✓  
*(consistent with the critical value method)*

#### Step 6 : 95% Confidence Interval

$$CI = \bar{x} \pm z_{\alpha/2} \cdot \frac{\sigma}{\sqrt{n}}$$

If μ₀ falls outside this interval, it is another way to conclude rejection of H₀.

In [6]:
moe1 = zcrit1 * (sig1 / np.sqrt(n1))   # margin of error
lo1  = xbar1 - moe1                     # lower bound
hi1  = xbar1 + moe1                     # upper bound
(lo1, hi1)

(np.float64(10.022819235130065), np.float64(10.577180764869937))

#### ✅ Final Conclusion - Q1

| Metric | Value |
|---|---|
| z-statistic | 2.12 |
| Critical value (±) | 1.96 |
| p-value | 0.034 |
| 95% CI | (10.02, 10.58) |

Both methods agree. μ₀ = 10 mm lies **outside** the confidence interval, |z| exceeds the critical threshold, and p-value < α.

**We reject H₀.** There is sufficient evidence that the true mean bolt diameter is not 10 mm. The factory's claim is questionable at the 5% significance level.

---
## 📌 Question 2 : One-Sample Proportion Z-Test

A survey of **800 voters** finds that **480 support** a new policy.  
A pollster claims that **55% of the population** supports the policy.

| Parameter | Value |
|---|---|
| Sample size (n) | 800 |
| Successes (x) | 480 |
| Hypothesized proportion (p₀) | 0.55 |
| Significance level (α) | 0.05 |

Perform a **two-tailed one-sample proportion Z-test**. Also compute the **95% CI**.

#### Step 1 : Hypotheses

- **H₀:** p = 0.55
- **H₁:** p ≠ 0.55 (two-tailed, the observed proportion could be higher or lower)

#### Step 2 : Why Proportion Z-Test?

- Data is count-based (successes out of n) → proportion test
- n is large → CLT applies, normal approximation is valid
- The standard error under H₀ uses p₀ (not p̂), because we assume H₀ is true when testing

In [7]:
# Q2 - given values
x2   = 480    # number of supporters
n2   = 800    # total surveyed
p02  = 0.55   # claimed population proportion
a2   = 0.05   # significance level

phat2 = x2 / n2   # observed sample proportion
phat2

0.6

#### Step 3 : Test Statistic

$$z = \frac{\hat{p} - p_0}{\sqrt{\frac{p_0(1-p_0)}{n}}}$$

The SE uses p₀ (not p̂) because under H₀ we assume the true proportion is p₀.

In [8]:
se_h02 = (p02 * (1 - p02) / n2) ** 0.5   # SE under null hypothesis
z2     = (phat2 - p02) / se_h02           # z-statistic
z2

2.8426762180748018

#### Step 4 : Critical Value Method

In [9]:
zcrit2 = norm.ppf(1 - a2/2)   # two-tailed critical value
zcrit2

np.float64(1.959963984540054)

**Decision:** |z| = 2.84 > z_critical = 1.96 → **Reject H₀**

#### Step 5 : p-value Method

In [10]:
pval2 = 2 * norm.sf(abs(z2))   # two-tailed p-value
pval2

np.float64(0.004473649256611822)

**Decision:** p-value = 0.0044 < α = 0.05 → **Reject H₀** ✓

#### Step 6 : 95% Confidence Interval

$$CI = \hat{p} \pm z_{\alpha/2} \cdot \sqrt{\frac{\hat{p}(1-\hat{p})}{n}}$$

**Note:** For the CI, the SE switches to p̂ in the formula. At this point H₀ is no longer assumed, so the estimate uses the observed proportion instead of p₀.

In [11]:
se_est2 = (phat2 * (1 - phat2) / n2) ** 0.5   # SE using p̂ (for CI)
moe2    = zcrit2 * se_est2                      # margin of error
lo2     = phat2 - moe2                          # lower bound
hi2     = phat2 + moe2                          # upper bound
(lo2, hi2)

(np.float64(0.5660524279777148), np.float64(0.6339475720222851))

#### ✅ Final Conclusion - Q2

| Metric | Value |
|---|---|
| Sample proportion (p̂) | 0.60 |
| z-statistic | 2.84 |
| Critical value (±) | 1.96 |
| p-value | 0.0044 |
| 95% CI | (0.566, 0.634) |

The claimed proportion p₀ = 0.55 lies **outside** the 95% CI, and the p-value is well below 0.05.

**We reject H₀.** The data suggests the true support level is significantly higher than the pollster's claimed 55%.

---
## 📌 Question 3 : Two-Proportion Z-Test (A/B Test Style)

| Group | n | Conversions |
|---|---|---|
| Control | 5000 | 600 |
| Treatment | 5000 | 680 |

Test at **α = 0.01** whether the treatment **significantly improves** the conversion rate.  
Also compute the **99% CI** for the difference in proportions.

> **Key concept:** When to use pooled vs unpooled SE?

#### Step 1 : Hypotheses

- **H₀:** p_ctrl = p_trt (treatment has no effect)
- **H₁:** p_trt > p_ctrl (treatment improves conversion, one-tailed)

One-tailed because we specifically want to know whether the treatment is **better**, not just different.

#### Step 2 : Pooled vs Unpooled SE

| Context | SE to use | Reason |
|---|---|---|
| **Hypothesis test** | Pooled | Under H₀ both groups share the same p, so merging gives a better common estimate |
| **Confidence interval** | Unpooled | H₀ is no longer assumed, each group uses its own p̂ |

In [12]:
# Q3 - given values
nc, xc = 5000, 600   # control:   sample size, conversions
nt, xt = 5000, 680   # treatment: sample size, conversions
a3     = 0.01        # significance level

# individual sample proportions
pc = xc / nc   # control proportion
pt = xt / nt   # treatment proportion
pc, pt

(0.12, 0.136)

#### Step 3 : Pooled Proportion & SE (for hypothesis test)

$$\hat{p}_{pool} = \frac{x_c + x_t}{n_c + n_t}$$

In [13]:
pp      = (xc + xt) / (nc + nt)                    # pooled proportion, assumes H₀: pc = pt
se_pool = (pp * (1 - pp) * (1/nc + 1/nt)) ** 0.5  # pooled SE
z3      = (pt - pc) / se_pool                       # z-statistic
pp, se_pool, z3

(0.128, 0.006681796165702753, 2.3945657130528804)

#### Step 4 : Critical Value (one-tailed, α = 0.01)

One-tailed → rejection region is only on the right side.  
We need z such that P(Z < z) = 0.99.

In [14]:
zcrit3 = norm.ppf(1 - a3)   # one-tailed critical value
zcrit3

np.float64(2.3263478740408408)

**Decision:** z = 2.39 > z_critical = 2.33 → **Reject H₀**

#### Step 5 : p-value (one-tailed)

In [15]:
pval3 = norm.sf(z3)   # P(Z > z), one-tailed
pval3

np.float64(0.008320030697044572)

**Decision:** p = 0.0083 < α = 0.01 → **Reject H₀** ✓

#### Step 6 : 99% Confidence Interval (Unpooled SE)

$$CI = (\hat{p}_t - \hat{p}_c) \pm z_{\alpha/2} \cdot \sqrt{\frac{\hat{p}_c(1-\hat{p}_c)}{n_c} + \frac{\hat{p}_t(1-\hat{p}_t)}{n_t}}$$

Unpooled SE is used here because the goal is to estimate the actual difference between the two proportions, not assume they are equal.

In [16]:
se_unpool  = np.sqrt(pc*(1-pc)/nc + pt*(1-pt)/nt)   # unpooled SE, each group uses its own proportion
zcrit3_ci  = norm.ppf(1 - a3/2)                     # two-tailed critical value for 99% CI
diff3      = pt - pc                                 # observed difference in proportions
lo3        = diff3 - zcrit3_ci * se_unpool           # lower bound
hi3        = diff3 + zcrit3_ci * se_unpool           # upper bound
(lo3, hi3)

(np.float64(-0.0012062312623899972), np.float64(0.03320623126239003))

#### ✅ Final Conclusion - Q3

| Metric | Value |
|---|---|
| Control p̂ | 0.12 (12%) |
| Treatment p̂ | 0.136 (13.6%) |
| z-statistic (pooled SE) | 2.39 |
| Critical value (one-tailed, α=0.01) | 2.33 |
| p-value | 0.0083 |
| 99% CI (unpooled SE) | (−0.0012, 0.0332) |

**We reject H₀.** The treatment group shows a statistically significant improvement in conversion rate at α = 0.01.

**Note on the CI vs the test:** The 99% CI barely includes 0 (lower bound ≈ −0.0012). This is because the CI uses a two-tailed z (2.576) while the hypothesis test was one-tailed (2.326). The two are answering slightly different questions, both are valid, but comparing them directly can be misleading.

---
**Pooled vs Unpooled - Quick Reference**

- **Pooled SE** → hypothesis test only. Under H₀, the groups share the same underlying proportion, so merging gives a better estimate.
- **Unpooled SE** → confidence intervals. H₀ is no longer assumed; each group's p̂ is used independently.

---
## 📌 Question 4 : Two-Sample Z-Test + Power Analysis

Drug A and Drug B are compared for blood pressure (BP) reduction.

| Parameter | Drug A | Drug B |
|---|---|---|
| Mean BP reduction (μ) | 12 mmHg | 10.5 mmHg |
| Population std dev (σ) | 4 mmHg | 3.5 mmHg |
| Sample size (n) | 100 | 120 |
| Significance level (α) | 0.05 | N/A |

**(a)** Test H₀: μ_A = μ_B at α = 0.05 using a two-sample Z-test.  
**(b)** Compute the power of this test, assuming the true difference is 2 mmHg.  
**(c)** Find the sample size (equal groups) required to achieve 90% power for a 2 mmHg difference.

#### Step 1 : Hypotheses

- **H₀:** μ_A = μ_B (both drugs reduce BP by the same amount)
- **H₁:** μ_A ≠ μ_B (two-tailed, either drug could be more effective)

#### Step 2 : Why Two-Sample Z-Test?

- Both population σ values are **known** → Z-test (unknown σ would require a t-test)
- Both sample sizes are large (100 and 120), CLT applies comfortably

In [17]:
# Q4 - given values
mu_a  = 12      # mean BP reduction, Drug A (mmHg)
mu_b  = 10.5    # mean BP reduction, Drug B (mmHg)
sig_a = 4       # population std dev, Drug A (mmHg)
sig_b = 3.5     # population std dev, Drug B (mmHg)
n_a   = 100     # sample size, Drug A
n_b   = 120     # sample size, Drug B
alpha = 0.05    # significance level

#### Step 3 : Standard Error and Test Statistic

$$z = \frac{(\bar{x}_A - \bar{x}_B) - 0}{\sqrt{\frac{\sigma_A^2}{n_A} + \frac{\sigma_B^2}{n_B}}}$$

The SE (denominator) quantifies how much the observed difference in means varies due to chance alone.

In [18]:
std_err = np.sqrt(sig_a**2 / n_a + sig_b**2 / n_b)   # standard error of the difference
z_stat  = (mu_a - mu_b) / std_err                     # z-statistic
std_err, z_stat

(np.float64(0.5119407517802557), np.float64(2.9300265602685536))

#### Step 4 : Critical Value Method

Two-tailed at α = 0.05 → 2.5% rejection region on each side.  
**Decision rule:** Reject H₀ if |z_stat| > z_critical.

In [19]:
z_crit = norm.ppf(1 - alpha / 2)   # two-tailed critical value
z_crit

np.float64(1.959963984540054)

**Decision:** |z| = 2.93 > z_critical = 1.96 → **Reject H₀**

#### Step 5 : p-value

Two-tailed p-value: probability of observing a z this extreme (or more) under H₀.

In [20]:
pval = 2 * norm.sf(abs(z_stat))   # two-tailed p-value
pval

np.float64(0.0033893303253593548)

**Decision:** p = 0.0034 < α = 0.05 → **Reject H₀** ✓  
Both methods agree, the observed 1.5 mmHg difference is statistically significant.

#### ✅ Part (a) Conclusion

| Metric | Value |
|---|---|
| Observed difference | 1.5 mmHg |
| Standard error | 0.512 |
| z-statistic | 2.93 |
| Critical value (±) | 1.96 |
| p-value | 0.0034 |

**We reject H₀.** Drug A shows a significantly greater BP reduction than Drug B at the 5% level.

---
#### Part (b) : Computing Power

**Power** is the probability that the test correctly rejects H₀ when H₁ is actually true.

Assume the true difference really is **2 mmHg**. Under H₁, the Z-statistic is no longer centered at 0. It shifts to Δ/SE, which is the non-centrality parameter. Power is the area of that shifted distribution that falls in the rejection region.

In [21]:
true_diff = 2   # assumed true difference in means (mmHg)

# non-centrality parameter, how many SEs away is the true difference from 0
ncp = true_diff / std_err
ncp

np.float64(3.906702080358072)

The true distribution is centered at z ≈ 3.91 (not 0), while the cutoff remains at 1.96.  
Most of the distribution already lies past the cutoff, so power should be high.

In [22]:
right_tail = norm.sf(z_crit - ncp)    # main contributor to power
left_tail  = norm.cdf(-z_crit - ncp)  # negligible, but included for correctness
power_b    = right_tail + left_tail    # total power
power_b

np.float64(0.974216931277533)

#### ✅ Part (b) Conclusion

**Power ≈ 97.4%.** With n = 100 and n = 120, the test has a very high chance of detecting a true 2 mmHg difference.

- z_β = ncp − z_critical = 3.91 − 1.96 = **1.95** → β ≈ 2.6%
- The test would miss the real effect only ~2.6% of the time

---
#### Part (c) : Required Sample Size for 90% Power

Fixing power = 90% and solving for n (equal group sizes):

$$n = \frac{(\sigma_A^2 + \sigma_B^2)(z_{\alpha/2} + z_{\beta})^2}{\Delta^2}$$

**Why add z_α/2 and z_β?** They measure the distance from the null mean to the cutoff, and from the cutoff to the true mean, respectively. Together they span the full separation the two distributions need.

In [23]:
beta = 0.10   # 1 - target power (0.90)

z_a = norm.ppf(1 - alpha / 2)   # critical value for α
z_b = norm.ppf(1 - beta)        # z-value for target power (90%)
z_a, z_b

(np.float64(1.959963984540054), np.float64(1.2815515655446004))

In [24]:
n_req   = (z_a + z_b)**2 * (sig_a**2 + sig_b**2) / true_diff**2   # required n (continuous)
n_final = math.ceil(n_req)                                          # always round up
n_final

75

#### ✅ Part (c) Conclusion

| Component | Value |
|---|---|
| z_α/2 (two-tailed, α = 0.05) | 1.96 |
| z_β (for 90% power) | 1.28 |
| z_α/2 + z_β | 3.24 |
| σ_A² + σ_B² | 28.25 |
| Δ (effect to detect) | 2 mmHg |
| **Required n per group** | **75** |

Only **n = 75 per group** is needed, which is less than the 100 and 120 in the original study. The original design was over-powered for this effect size, which is why Part (b) yielded 97.4%, well above the 90% target.

---
## 📌 Question 5 : Multiple Testing Corrections

Three independent A/B tests yield raw p-values: **[0.03, 0.04, 0.048]**.

**(a)** Apply **Bonferroni correction** at α = 0.05. Which hypotheses survive?  
**(b)** Apply **Benjamini–Hochberg (FDR) correction** at α = 0.05. Which survive?  
**(c)** Explain the trade-off between FWER control (Bonferroni) and FDR control (BH). When would you choose each in practice?

In [25]:
from statsmodels.stats.multitest import multipletests

#### Step 1 : Setup

In [26]:
p_vals = [0.03, 0.04, 0.048]   # raw p-values from three tests
alpha  = 0.05                   # family-wise significance level
n      = len(p_vals)            # number of tests

---
#### Part (a) : Bonferroni Correction

**Idea:** Divide α equally across all m tests.  
$$\alpha_{Bonferroni} = \frac{\alpha}{m}$$

Reject H₀ᵢ if (pᵢ < α/m) or (pᵢ*m < α).  
This controls the **Family-Wise Error Rate (FWER)**, the probability of making *any* false positive across all tests.

In [27]:
alpha_bonf = alpha / n   # adjusted threshold per test
print(f"Bonferroni threshold: {alpha_bonf:.4f}\n")

for i, p in enumerate(p_vals, start=1):
    decision = "Reject H₀" if p < alpha_bonf else "Fail to reject H₀"
    print(f"  H{i}: p = {p:.3f}  {'<' if p < alpha_bonf else '>'}  {alpha_bonf:.4f}  →  {decision}")

Bonferroni threshold: 0.0167

  H1: p = 0.030  >  0.0167  →  Fail to reject H₀
  H2: p = 0.040  >  0.0167  →  Fail to reject H₀
  H3: p = 0.048  >  0.0167  →  Fail to reject H₀


In [28]:
reject_bonf, pvals_bonf, _, _ = multipletests(
    p_vals, alpha=alpha, method='bonferroni'
)

print("Bonferroni corrected p-values:", pvals_bonf)
print("Reject (Bonferroni):", reject_bonf)

Bonferroni corrected p-values: [0.09  0.12  0.144]
Reject (Bonferroni): [False False False]


---
#### Part (b) : Benjamini–Hochberg (FDR) Correction

**Procedure:**
1. Sort p-values in ascending order: p_(1) ≤ p_(2) ≤ … ≤ p_(m)
2. For each rank i, compute the BH threshold: (i / m) × α
3. Find the **largest** i where p_(i) < threshold (call it k)
4. Reject all H₀ᵢ for i ≤ k

This controls the **False Discovery Rate (FDR)**, the expected proportion of false positives among rejected tests.

In [29]:
sorted_p = sorted(enumerate(p_vals, start=1), key=lambda x: x[1])
print(f"{'Rank':<6} {'Orig. idx':<12} {'p-value':<10} {'BH threshold':<15} {'Decision'}")
print("-" * 60)

k = 0
for rank, (orig_idx, p) in enumerate(sorted_p, start=1):
    threshold = (rank / n) * alpha
    if p < threshold:
        k = rank

for rank, (orig_idx, p) in enumerate(sorted_p, start=1):
    threshold = (rank / n) * alpha
    decision  = "Reject H₀" if rank <= k else "Fail to reject H₀"
    marker    = " <- cutoff (k)" if rank == k else ""
    print(f"{rank:<6} H{orig_idx:<11} {p:<10.3f} {threshold:<15.4f} {decision}{marker}")

print(f"\nLargest significant rank k = {k}")
print(f"Reject all hypotheses with rank <= {k}")

Rank   Orig. idx    p-value    BH threshold    Decision
------------------------------------------------------------
1      H1           0.030      0.0167          Reject H₀
2      H2           0.040      0.0333          Reject H₀
3      H3           0.048      0.0500          Reject H₀ <- cutoff (k)

Largest significant rank k = 3
Reject all hypotheses with rank <= 3


In [30]:
reject_bh, pvals_bh, _, _ = multipletests(
    p_vals, alpha=alpha, method='fdr_bh'
)

print("BH corrected p-values:", pvals_bh)
print("Reject (BH):", reject_bh)

BH corrected p-values: [0.048 0.048 0.048]
Reject (BH): [ True  True  True]


---
#### Part (c) : Bonferroni vs Benjamini–Hochberg

| | Bonferroni (FWER) | Benjamini–Hochberg (FDR) |
|---|---|---|
| **Controls** | Probability of ≥1 false positive | Expected proportion of false positives |
| **Strictness** | Very strict | Less strict |
| **False Positives** | Strongly minimized | Allows a small, bounded rate |
| **False Negatives** | Higher (misses real effects) | Lower (detects more real effects) |
| **Best used when** | High-stakes decisions where any false alarm is costly | Exploratory analysis where some false discoveries are acceptable |

**When to use Bonferroni:** Clinical trials, regulatory submissions, safety testing. In those contexts, a single false positive can cause real harm or waste serious resources.

**Go with BH when:** Genomics, large-scale A/B testing, exploratory studies. The goal there is to find as many real effects as possible, and a small fraction of false discoveries is acceptable as long as it stays bounded and gets followed up later.

---
## 📌 Question 6 : One-Sample Z-Test + Power Analysis

Weekly sales data over **52 weeks** shows a sample mean of **\$48,200**, with a known population σ = \$1,200. Last year's mean was **\$47,000**.

| Parameter | Value |
|---|---|
| Sample size (n) | 52 weeks |
| Sample mean (x̄) | \$48,200 |
| Population std dev (σ) | \$1,200 |
| Last year mean (μ₀) | \$47,000 |
| Significance level (α) | 0.01 |

**(a)** Test at α = 0.01 (one-tailed) whether sales have significantly increased.  
**(b)** Compute the required n for 90% power to detect a $500 improvement.  
**(c)** With current n = 52, what is the minimum detectable effect at 80% power?

#### Step 1 : Hypotheses

- **H₀:** μ = 47,000 (sales have not increased)
- **H₁:** μ > 47,000 (one-tailed, testing for an increase only)

#### Step 2 : Why Z-Test?

- Population σ is **known** → use Z-test (unknown σ would require a t-test)
- n = 52 is large enough for the Central Limit Theorem to apply
- One-tailed because the question asks specifically whether sales **increased**, not just changed

In [31]:
# Q6 - given values
n6    = 52       # number of weeks (sample size)
xbar6 = 48200    # sample mean weekly sales ($)
mu06  = 47000    # last year's mean, null hypothesis value ($)
sig6  = 1200     # known population std dev ($)
a6    = 0.01     # significance level

#### Step 3 : Test Statistic

$$z = \frac{\bar{x} - \mu_0}{\sigma / \sqrt{n}}$$

Measures how many standard errors the sample mean sits above the null mean.

In [32]:
z6 = (xbar6 - mu06) / (sig6 / n6**0.5)   # z-statistic
z6

7.211102550927978

#### Step 4 : Critical Value Method

One-tailed at α = 0.01 → entire rejection region on the right side.  
**Decision rule:** Reject H₀ if z_stat > z_critical.

In [33]:
zcrit6 = norm.ppf(1 - a6)   # one-tailed critical value at α = 0.01
zcrit6

np.float64(2.3263478740408408)

**Decision:** z = 7.21 > z_critical = 2.33 → **Reject H₀**

#### Step 5 : p-value Method

One-tailed p-value: probability of observing a z this large or larger under H₀.

In [34]:
pval6 = norm.sf(z6)                    # P(Z > z6), one-tailed
print(f"{pval6:.20f}")                # high precision to show how small it is

0.00000000000027750317


**Decision:** p-value << α = 0.01 → **Reject H₀** ✓

#### Step 6 : 99% Confidence Interval

$$CI = \bar{x} \pm z_{\alpha/2} \cdot \frac{\sigma}{\sqrt{n}}$$

For a one-tailed test at α = 0.01, the matching two-sided CI uses α/2 = 0.005 on each side.

In [35]:
zcrit6_ci = norm.ppf(1 - a6/2)              # two-tailed critical value for 99% CI
moe6      = zcrit6_ci * (sig6 / n6**0.5)    # margin of error
lo6       = xbar6 - moe6                     # lower bound
hi6       = xbar6 + moe6                     # upper bound
(lo6, hi6)

(np.float64(47771.356093963615), np.float64(48628.643906036385))

#### ✅ Part (a) Conclusion

| Metric | Value |
|---|---|
| z-statistic | 7.21 |
| Critical value (one-tailed) | 2.33 |
| p-value | ~6.2e-13 |
| 99% CI | ($47,771, $48,629) |

**We reject H₀.** μ₀ = $47,000 falls far outside the confidence interval, and the z-statistic is extreme. Sales have significantly increased at the 1% level.

---
#### Part (b) : Required Sample Size for 90% Power

Find n such that the test has 90% power to detect a true $500 increase.

$$n = \left(\frac{(z_{\alpha} + z_{\beta}) \cdot \sigma}{\Delta}\right)^2$$

In [36]:
power_b6 = 0.90            # target power
diff_b6  = 500             # effect to detect ($)
beta_b6  = 1 - power_b6   # type II error rate

z_a6 = norm.ppf(1 - a6)          # one-tailed critical value (α = 0.01)
z_b6 = norm.ppf(1 - beta_b6)     # z for target power (90%)
z_a6, z_b6

(np.float64(2.3263478740408408), np.float64(1.2815515655446004))

In [37]:
n_req6   = ((z_a6 + z_b6) * sig6 / diff_b6) ** 2   # required n (continuous)
n_final6 = math.ceil(n_req6)                          # always round up
n_final6

75

#### ✅ Part (b) Conclusion

| Component | Value |
|---|---|
| z_α (one-tailed, α = 0.01) | 2.33 |
| z_β (for 90% power) | 1.28 |
| σ | \$1,200 |
| Δ (effect to detect) | \$500 |
| **Required n** | **~65 weeks** |

Roughly 65 weeks of data are needed to reliably detect a $500 improvement. The current n = 52 is close but falls short for this smaller effect.

---
#### Part (c) : Minimum Detectable Effect at 80% Power

With n fixed at 52, find the smallest true difference Δ the test can detect at 80% power.

$$\Delta = \frac{(z_{\alpha} + z_{\beta}) \cdot \sigma}{\sqrt{n}}$$

In [38]:
n_c6      = 52                         # fixed sample size
power_c6  = 0.80                       # target power
beta_c6   = 1 - power_c6              # type II error rate
z_beta_c6 = norm.ppf(1 - beta_c6)    # z for 80% power
z_beta_c6

np.float64(0.8416212335729143)

In [39]:
mde6 = ((zcrit6 + z_beta_c6) * sig6) / (n_c6 ** 0.5)   # minimum detectable effect
math.ceil(mde6)

528

#### ✅ Part (c) Conclusion

| Component | Value |
|---|---|
| z_α (one-tailed, α = 0.01) | 2.33 |
| z_β (for 80% power) | 0.84 |
| σ | $1,200 |
| n | 52 |
| **Minimum detectable effect** | **~530** |

With n = 52, the test can reliably detect a true increase of roughly $530 or more at 80% power. Effects smaller than that are likely to go undetected.

---
## 📌 Question 7 : Effect Size vs Statistical Significance

A manufacturing plant compares defect rates between two production lines.

| Group | n | Defect rate |
|---|---|---|
| Line A | 4800 | 2.1% |
| Line B | 5200 | 1.7% |

**(a)** Test equality of proportions (two-tailed, α = 0.05).  
**(b)** Compute Cohen's h = 2·(arcsin(√p₁) − arcsin(√p₂)) and interpret it.  
**(c)** The test is statistically significant. Compute NNT = 1/|p₁ − p₂|.  
**(d)** Argue: when is statistical significance misleading without effect size?

#### Step 1 : Hypotheses

- **H₀:** p_A = p_B (defect rates are equal across both lines)
- **H₁:** p_A ≠ p_B (two-tailed, either line could have a higher defect rate)

#### Step 2 : Why Pooled Proportion?

H₀ asserts that both lines share the same underlying defect rate. Under that assumption, pooling the counts gives a better common estimate of p than using each line's proportion separately.

In [40]:
# Q7 - given values
p17   = 0.021   # Line A defect rate
n17   = 4800    # Line A sample size
p27   = 0.017   # Line B defect rate
n27   = 5200    # Line B sample size
a7    = 0.05    # significance level

#### Step 3 : Pooled Proportion & Test Statistic

$$\hat{p}_{pool} = \frac{x_1 + x_2}{n_1 + n_2}, \qquad z = \frac{\hat{p}_1 - \hat{p}_2}{\sqrt{\hat{p}_{pool}(1 - \hat{p}_{pool})\left(\frac{1}{n_1} + \frac{1}{n_2}\right)}}$$

In [41]:
x17 = p17 * n17   # Line A defect count (since p = x/n)
x27 = p27 * n27   # Line B defect count
x17, x27

(100.80000000000001, 88.4)

In [42]:
pp7  = (x17 + x27) / (n17 + n27)                        # pooled proportion
se7  = (pp7 * (1 - pp7) * (1/n17 + 1/n27)) ** 0.5      # pooled SE
z7   = (p17 - p27) / se7                                 # z-statistic
pp7, se7, z7

(0.018920000000000003, 0.0027270335860577846, 1.4667952827755315)

#### Step 4 : Critical Value Method

In [43]:
zcrit7 = norm.ppf(1 - a7/2)   # two-tailed critical value
zcrit7

np.float64(1.959963984540054)

#### Step 5 : p-value Method

In [44]:
pval7 = 2 * norm.sf(abs(z7))   # two-tailed p-value
pval7

np.float64(0.14243175333071545)

**Decision:** |z| < z_critical and p-value > α → **Fail to reject H₀**

Despite the difference in observed rates, there is not enough evidence at the 5% level to conclude the lines differ significantly.

#### ✅ Part (a) Conclusion

| Metric | Value |
|---|---|
| Line A defect rate | 2.1% |
| Line B defect rate | 1.7% |
| z-statistic | 1.47 |
| Critical value (±) | 1.96 |
| p-value | 0.142 |

**We fail to reject H₀.** The observed difference does not clear the significance threshold at α = 0.05.

---
#### Part (b) : Cohen's h (Effect Size)

Cohen's h measures the magnitude of the difference between two proportions on an arcsine-transformed scale. 
- Helps in determining if the difference is meaningful or not.
  
Benchmarks: h = 0.2 (small), h = 0.5 (medium), h = 0.8 (large).

In [45]:
h7 = 2 * (math.asin(math.sqrt(p17)) - math.asin(math.sqrt(p27)))   # Cohen's h
abs(h7)

0.029338989746430233

**Interpretation:** h ≈ 0.03 falls well below the 0.2 threshold for a small effect. The difference between the two lines is negligible in practical terms.

---
#### Part (c) : Number Needed to Treat (NNT)

NNT answers: how many units need to switch from Line A to Line B in order to prevent one additional defect? A smaller NNT means the intervention is more impactful.

In [46]:
nnt7 = 1 / abs(p17 - p27)   # number needed to treat
nnt7

250.0

**Interpretation:** NNT ≈ 250. Switching 250 units from Line A to Line B would, on average, prevent a single extra defect. At this scale, the operational cost of switching would likely outweigh the gain.

---
#### Part (d) : When Statistical Significance Is Misleading

Statistical significance does not equal practical importance. With very large sample sizes, even trivially small differences can produce a p-value below α. The test answers only one question: is the observed difference unlikely under H₀? It says nothing about whether the difference is large enough to matter.

In this case, the 0.4 percentage point gap between the lines has an NNT of 250 and a Cohen's h near zero. A sufficiently large sample would make this difference statistically significant, but the conclusion would be misleading without also reporting effect size. 

┌──────────────────────────────────────────────────────────────────────────┐
│  Q8                                                           │
│                                                                          │
│  A food delivery company claims the average delivery time is 30 minutes. │
│  The population standard deviation is known to be σ = 8 minutes.         │
│                                                                          │
│  You collect a sample of n = 100 deliveries and observe a mean of        │
│  x̄ = 31.5 minutes.                                                      │
│                                                                          │
│  (a) Perform a one-tailed Z-test at α = 0.05 to test whether delivery    │
│      time has increased. Compute the test statistic and conclusion.      │
│                                                                          │
│  (b) Compute the p-value and interpret it.                               │
│                                                                          │
│  (c) Suppose the true mean delivery time is actually 32 minutes.         │
│      Compute the power of the test.                                      │
│                                                                          │
│  (d) Compute the probability of Type II error (β).                       │
│                                                                          │
│  (e) What sample size is required to achieve 90% power to detect a       │
│      2-minute increase (from 30 to 32 minutes) at α = 0.05?              │
│                                                                          │
│  (f) Explain intuitively how sample size affects power and why this      │
│      matters in real-world A/B testing.                                  │
└──────────────────────────────────────────────────────────────────────────┘

## 📌 Question 8 : One-Sample Z-Test + Power Analysis

A food delivery company claims the average delivery time is 30 minutes. The population standard deviation is known to be σ = 8 minutes.

You collect a sample of n = 100 deliveries and observe a mean of x̄ = 31.5 minutes.

| Parameter | Value |
|---|---|
| Sample size (n) | 100 deliveries |
| Sample mean (x̄) | 31.5 minutes |
| Population std dev (σ) | 8 minutes |
| Claimed mean (μ₀) | 30 minutes |
| Significance level (α) | 0.05 |

**(a)** Perform a one-tailed Z-test at α = 0.05 to test whether delivery time has increased. Compute the test statistic and conclusion.  
**(b)** Compute the p-value and interpret it.  
**(c)** Suppose the true mean delivery time is actually 32 minutes. Compute the power of the test.  
**(d)** Compute the probability of Type II error (β).  
**(e)** What sample size is required to achieve 90% power to detect a 2-minute increase (from 30 to 32 minutes) at α = 0.05?  
**(f)** Explain intuitively how sample size affects power and why this matters in real-world A/B testing.

#### Step 1 : Hypotheses

- **H₀:** μ = 30 (delivery time has not increased)
- **H₁:** μ > 30 (one-tailed, testing for an increase only)

#### Step 2 : Why Z-Test?

- Population σ is **known** → use Z-test (unknown σ would require a t-test)
- n = 100 is large enough for the Central Limit Theorem to apply
- One-tailed because the question asks specifically whether delivery time **increased**, not just changed

#### Step 3 : Test Statistic

$$z = \frac{\bar{x} - \mu_0}{\sigma / \sqrt{n}}$$

Measures how many standard errors the sample mean sits above the null mean.

In [ ]:
# Q8 - given values
mu8    = 30      # claimed population mean (minutes)
sig8   = 8       # known population std dev (minutes)
xbar8  = 31.5    # observed sample mean (minutes)
n8     = 100     # sample size
a8     = 0.05    # significance level

se8 = sig8 / n8**0.5          # standard error of the mean
z8  = (xbar8 - mu8) / se8     # z-statistic
z8

#### Step 4 : Critical Value Method

One-tailed at α = 0.05 → entire rejection region on the right side.  
**Decision rule:** Reject H₀ if z_stat > z_critical.

In [ ]:
zcrit8 = norm.ppf(1 - a8)   # one-tailed critical value at α = 0.05
zcrit8

**Decision:** z = 1.875 > z_critical = 1.645 → **Reject H₀**

#### Step 5 : p-value Method

One-tailed p-value: probability of observing a z this large or larger under H₀.

In [ ]:
pval8 = norm.sf(z8)   # P(Z > z8), one-tailed
pval8

**Decision:** p-value ≈ 0.030 < α = 0.05 → **Reject H₀**

#### Step 6 : 95% Confidence Interval

$$CI = \bar{x} \pm z_{\alpha/2} \cdot \frac{\sigma}{\sqrt{n}}$$

For a one-tailed test at α = 0.05, the matching two-sided CI uses α/2 = 0.025 on each side.

In [ ]:
zcrit8_ci = norm.ppf(1 - a8/2)           # two-tailed critical value for 95% CI
moe8      = zcrit8_ci * (sig8 / n8**0.5)  # margin of error
lo8       = xbar8 - moe8                   # lower bound
hi8       = xbar8 + moe8                   # upper bound
(lo8, hi8)

#### ✅ Part (a) Conclusion

| Metric | Value |
|---|---|
| z-statistic | 1.875 |
| Critical value (one-tailed) | 1.645 |
| p-value | ~0.030 |
| 95% CI | (29.93, 33.07) |

**We reject H₀.** The sample mean of 31.5 minutes is significantly above the claimed 30 minutes at the 5% level. There is sufficient evidence that delivery times have increased.

---
#### Part (b) : p-value Interpretation

The p-value of ~0.030 means: if the true mean really were 30 minutes, there would be about a 3% chance of observing a sample mean of 31.5 or higher just by chance with n = 100. Since 3% < 5%, this result is unlikely under H₀ and the null is rejected.

---
#### Part (c) : Power of the Test (True μ = 32)

Power is the probability of correctly rejecting H₀ when the true mean is 32 minutes. The critical value in original units defines the rejection boundary. Power is the area to the right of that boundary under the true distribution centered at 32.

In [ ]:
mu_true8  = 32                            # assumed true population mean
diff8     = mu_true8 - mu8                 # true effect size (minutes)
z_eff8    = diff8 / se8                    # effect in standard error units
z_shift8  = zcrit8 - z_eff8               # critical z relative to the true center
power8    = norm.sf(z_shift8)              # P(reject H0 | true mean = 32)
power8

#### ✅ Part (c) Conclusion

| Component | Value |
|---|---|
| True mean (μ) | 32 minutes |
| Effect size (Δ) | 2 minutes |
| z_critical | 1.645 |
| z_shift | ~0.395 |
| **Power** | **~34.6%** |

With n = 100, the test has only about 35% power to detect a true 2-minute increase. That is quite low, meaning most of the time this test would miss the effect.

---
#### Part (d) : Type II Error (β)

β is the probability of failing to reject H₀ when the alternative is true. It is the complement of power.

In [ ]:
beta8 = 1 - power8   # probability of Type II error
beta8

#### ✅ Part (d) Conclusion

| Metric | Value |
|---|---|
| Power | ~34.6% |
| **β (Type II error)** | **~65.4%** |

There is roughly a 65% chance the test fails to detect the true 2-minute increase. A β this high is unacceptable in most practical settings.

---
#### Part (e) : Required Sample Size for 90% Power

Find n such that the test has 90% power to detect a true 2-minute increase at α = 0.05.

$$n = \left(\frac{(z_{\alpha} + z_{\beta}) \cdot \sigma}{\Delta}\right)^2$$

In [ ]:
power_e8  = 0.90                          # target power
beta_e8   = 1 - power_e8                   # acceptable Type II error rate
z_a8      = norm.ppf(1 - a8)               # one-tailed critical value (α = 0.05)
z_b8      = norm.ppf(1 - beta_e8)          # z for 90% power
z_a8, z_b8

In [ ]:
n_req8   = ((z_a8 + z_b8) * sig8 / diff8) ** 2   # required n (continuous)
n_final8 = math.ceil(n_req8)                        # always round up
n_final8

#### ✅ Part (e) Conclusion

| Component | Value |
|---|---|
| z_α (one-tailed, α = 0.05) | 1.645 |
| z_β (for 90% power) | 1.282 |
| σ | 8 minutes |
| Δ (effect to detect) | 2 minutes |
| **Required n** | **~263 deliveries** |

About 263 deliveries are needed to reliably detect a 2-minute increase with 90% power. The original n = 100 falls well short.

---
#### Part (f) : Sample Size, Power, and A/B Testing

A larger sample shrinks the standard error, pulling the sampling distributions for H₀ and H₁ further apart. The overlap between those two distributions is exactly what determines power. Less overlap means the test is more likely to detect a real effect.

In A/B testing, running an underpowered experiment is a common and costly mistake. If the sample is too small, a genuine improvement (faster delivery, higher conversion) goes undetected, and the null is retained when it should be rejected. The result is a false negative, and a real effect gets ignored.

The fix is to compute the required sample size before the experiment starts, not after. Determining n upfront from a target power (usually 80% or 90%) and a minimum detectable effect ensures the test has a realistic chance of catching what it is looking for.